lib/engine/game/g_18_oe/step/buy_sell_par_shares.rb

```diff
@@ -1,38 +1,102 @@
 
-            actions << 'pass' if !can_float_minor?(entity) && !bought? && !actions.empty?
+            actions << 'pass' if !can_float_minor?(entity) && !actions.empty?
             actions
           end
 
+          def corporation_actions(entity)
+            if @round.current_actions[0].instance_of?(Engine::Action::BuyShares) &&
+              @round.current_actions[0].bundle.corporation != entity
+              return []
+            end
+            return [] unless can_convert?(entity)
+
+            %w[convert pass]
+          end
+
           def can_buy_any_from_ipo?(entity)
             return unless @game.corporations.all?(&:ipoed)
 
             super
           end
 
+          def can_buy?(entity, bundle)
+            if @round.current_actions[0].instance_of?(Engine::Action::Convert) &&
+              @round.current_actions[0].entity != bundle.corporation
+              return false
+            end
+
+            super
+          end
+
+          def can_convert?(corporation)
+            if !@game.corporations.all?(&:ipoed) || @round.current_actions.any? { |x| x.instance_of?(Engine::Action::Convert) }
+              return
+            end
+
+            return if @round.current_actions[0].instance_of?(Engine::Action::BuyShares) &&
+              (@round.current_actions[0].bundle.corporation != corporation || corporation.share_holders[current_entity] <= 50)
+
+            corporation.total_shares == 4 && corporation.share_holders.include?(current_entity) &&
+              corporation.share_holders[current_entity] >= 50
+          end
+
+          def can_convert_any?(entity)
+            return if @round.current_actions.any? { |x| x.instance_of?(Engine::Action::Convert) }
+
+            entity.shares_by_corporation.reject { |c| c.type == :minor }.any? { |c| can_convert?(c[0]) }
+          end
+
           def can_float_minor?(entity)
             return unless entity.player?
 
             !bought? && entity.companies.any? { |company| @game.company_becomes_minor?(company) }
           end
 
+          def float_major(corporation)
+            shares = corporation.share_holders.keys.flat_map { |share| share.shares_of(corporation) }
+
+            shares.each { |share| share.percent = share.president ? 20 : 10 }
+            6.times do |index|
+              share = Share.new(corporation, owner: corporation.ipo_owner, percent: 10, index: 4 + index)
+              corporation.ipo_owner.shares_by_corporation[corporation] << share
+            end
+
+            # Set corporation type to :major
+            corporation.type = 'major'
+
+            # Majors are affected by the stock market, set tokens in the correct place
+            @game.stock_market.move_right(corporation)
+            @game.stock_market.move_right(corporation)
+            @game.stock_market.move_up(corporation)
+            corporation.tokens += [40, 60, 60, 80, 80, 80].map { |price| Engine::Token.new(corporation, price: price) }
```

Comment:
corporation.tokens.concat() and then you don't need to make tokens an attr_accessor in operator.rb